### PR TITLE
[NO MERGE] Remove LookupByKey and Fetch nodes from TransactionCommitter

### PR DESCRIPTION
**Pull request by tudor-da**
This *draft* PR is meant for assessing the impact that _slimmed-down_ transaction trees have downstream (after the execution engine, especially on the read-path of the participant) in terms of throughput and other related metrics. A complete description can be found in [the JIRA task](https://digitalasset.atlassian.net/jira/software/projects/KVL/boards/494?selectedIssue=KVL-627)

As discussed with @gerolf-da and @meiersi-da , the removal of `LookupByKey` and `Fetch` are impending the contract divulgence computation, hence the failing tests. However, the benchmarked scenario is not impacted by this.

For means of correctness checking, there's a new test case in `TransactionCommitterSpec`

CHANGELOG_BEGIN
CHANGELOG_END

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 1]
```diff
@@ -252,9 +254,57 @@ private[kvutils] class TransactionCommitter(
...
     (commitContext, transactionEntry) => {
       val blindingInfo = Blinding.blind(transactionEntry.transaction)
-      buildFinalResult(commitContext, transactionEntry, blindingInfo)
+      buildFinalResultState(commitContext, transactionEntry, blindingInfo)
     }
 
+  private def removeUnnecessaryNodes: Step = (commitContext, transactionEntry) => {
+    val transaction = transactionEntry.submission.getTransaction
+    val nodes: Map[String, TransactionOuterClass.Node] =
+      transaction.getNodesList.asScala.map(n => n.getNodeId -> n)(scala.collection.breakOut)
```
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:263]:
```suggestion
      transaction.getNodesList.iterator.asScala.map(n => n.getNodeId -> n).toMap
```
I appreciate the intent of not creating intermediate collections, but I would recommend using iterators or views in this regard. The hope is to make it easier on us to migrate to 2.13 (one day), where `CanBuildFrom` no longer exists.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 2]
```diff
@@ -252,9 +254,57 @@ private[kvutils] class TransactionCommitter(
...
+    val nodesToKeep: Set[String] = nodes.collect {
+      case (id, node) if node.hasCreate || node.hasExercise => id
+    }(scala.collection.breakOut)
```
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:266]:
```suggestion
    val nodesToKeep: Set[String] = nodes.iterator.collect {
      case (id, node) if node.hasCreate || node.hasExercise => id
    }.toSet
```
Same point as above.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 3]
```diff
@@ -252,9 +254,57 @@ private[kvutils] class TransactionCommitter(
...
+
+    val roots = transaction.getRootsList.asScala.filter(nodesToKeep)
+
+    val stripUnnecessaryChildren: TransactionOuterClass.Node => TransactionOuterClass.Node = {
```
Comment (stefanobaghino-da) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:270]: If you make a `def` of this I'm relatively sure the Scala compiler will turn it into a private method of the class. I'm not sure of what happens to function objects. Would this be compiled to a method or would this cause the function object to be created on every call of `removeUnnecessaryNodes`.

**Review by stefanobaghino-da — COMMENTED** [review]
I understand this is still a draft, so just some early feedback. Looks good so far. 🙂 

For my understanding: are we planning to make this change part of the protocol between the committer and the indexer, allowing the indexer to assume there will be no unnecessary¹ nodes in the transaction? If so, do we ultimately want to also remove the filtering we're already performing there in this PR?

¹ from a Ledger API perspective

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 3 above)
Comment (stefanobaghino-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:270]: Looks like on this toy experiment

```scala
object Foo {

  def bar(ns: List[Int]): List[Int] = {
    val predicate: Int => Boolean = _ < 5
    ns.filter(predicate)
  }

}
```

the compiler actually compiles the function to a private method (`$anonfun$bar$1`)

```
> javap -c Foo
Compiled from "Foo.scala"
public final class Foo {
  public static scala.collection.immutable.List<java.lang.Object> bar(scala.collection.immutable.List<java.lang.Object>);
    Code:
       0: getstatic     #18                 // Field Foo$.MODULE$:LFoo$;
       3: aload_0
       4: invokevirtual #20                 // Method Foo$.bar:(Lscala/collection/immutable/List;)Lscala/collection/immutable/List;
       7: areturn
}
> javap -c Foo$
Compiled from "Foo.scala"
public final class Foo$ {
  public static final Foo$ MODULE$;

  public static {};
    Code:
       0: new           #2                  // class Foo$
       3: dup
       4: invokespecial #17                 // Method "<init>":()V
       7: putstatic     #19                 // Field MODULE$:LFoo$;
      10: return

  public scala.collection.immutable.List<java.lang.Object> bar(scala.collection.immutable.List<java.lang.Object>);
    Code:
       0: invokedynamic #41,  0             // InvokeDynamic #0:apply$mcZI$sp:()Lscala/runtime/java8/JFunction1$mcZI$sp;
       5: astore_2
       6: aload_1
       7: aload_2
       8: invokevirtual #47                 // Method scala/collection/immutable/List.filter:(Lscala/Function1;)Lscala/collection/immutable/List;
      11: areturn

  public static final boolean $anonfun$bar$1(int);
    Code:
       0: iload_0
       1: iconst_5
       2: if_icmpge     9
       5: iconst_1
       6: goto          10
       9: iconst_0
      10: ireturn
}
```

Smart compiler. 🙂

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 4]
```diff
@@ -252,9 +254,57 @@ private[kvutils] class TransactionCommitter(
...
+
+    val newNodes = nodes
+      .collect {
+        case (id, node) if nodesToKeep(id) =>
+          stripUnnecessaryChildren(node)
+      }
+
+    val newTx = transaction
+      .newBuilderForType()
+      .addAllRoots(newRoots.asJavaCollection)
```
Comment (dajmaki) [inline ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:295]: Transaction version is lost here. Might be better to clear roots and nodes and add them back instead so we don't forget other fields (and new ones might be added).

Comment (tudor-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:295]: Thanks @dajmaki . I noticed, I've added it back.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 5]
```diff
@@ -67,7 +68,8 @@ private[kvutils] class TransactionCommitter(
     "validate_ledger_time" -> validateLedgerTime,
     "validate_contract_keys" -> validateContractKeys,
     "validate_model_conformance" -> validateModelConformance,
-    "blind" -> blind
+    "blind" -> blind,
+    "remove_unnecessary_nodes" -> removeUnnecessaryNodes
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:72]: Hmm ok 😄

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 6]
```diff
@@ -252,9 +254,53 @@ private[kvutils] class TransactionCommitter(
...
     }
 
+
+
+  private[committer] def removeUnnecessaryNodes: Step = (commitContext, transactionEntry) => {
+    val transaction = transactionEntry.submission.getTransaction
+    val nodes = transaction.getNodesList.asScala
+    val nodesToKeep = nodes.collect {
+      case node if node.hasCreate || node.hasExercise => node.getNodeId
+    }.toSet
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:267]: That could have worked if all the filters were based on the same input. But the children list in nodes consists of ids, so filtering those needs additional context.

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:72]: We usually prefer to have dangling commas to make future diffs more succinct.

There's a `scalafmt` rule for that (that would be preferable to leave it to reviews), but unfortunately that would need to happen in a moment where very few people have open PRs to minimize conflicts (so in the end we never did it).

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 6 above)
[the remark this answers is not in the document]
Comment (stefanobaghino-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:267]: But the point on the intermediate collection is still valid. I was suggesting to use `iterator` in https://github.com/digital-asset/daml/pull/7743#discussion_r508298354 to that end (where you originally used `breakOut`, which will go away in Scala 2.13).

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 7]
```diff
@@ -252,9 +254,53 @@ private[kvutils] class TransactionCommitter(
...
+      } else node
+
+    val newNodes = nodes
+      .collect {
+        case node if nodesToKeep(node.getNodeId) => stripUnnecessaryNodes(node)
+      }
+
+    val newTx = transaction
+      .newBuilderForType()
+      .addAllRoots(roots.filter(nodesToKeep).asJavaCollection)
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala]: Good catch, thanks

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala [hunk 8]
```diff
@@ -252,9 +254,53 @@ private[kvutils] class TransactionCommitter(
   private def blind: Step =
     (commitContext, transactionEntry) => {
       val blindingInfo = Blinding.blind(transactionEntry.transaction)
-      buildFinalResult(commitContext, transactionEntry, blindingInfo)
+      buildFinalResultState(commitContext, transactionEntry, blindingInfo)
     }
 
+
+
+  private[committer] def removeUnnecessaryNodes: Step = (commitContext, transactionEntry) => {
```
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:263]: Added a comment for this method

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 5 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala]: Didn't consider that, thanks for the clarification

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (fabiotudone-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:263]: Thanks, can you also expand it a bit to explain why we're doing that?

**Comment by stefanobaghino-da** [comment]
> I understand this is still a draft, so just some early feedback. Looks good so far. 
> 
> For my understanding: are we planning to make this change part of the protocol between the committer and the indexer, allowing the indexer to assume there will be no unnecessary¹ nodes in the transaction? If so, do we ultimately want to also remove the filtering we're already performing there in this PR?
> 
> ¹ from a Ledger API perspective

In my review I left this comment but it probably went unnoticed. tl;dr "do we ultimately want to also remove the filtering we're already performing [in the indexer] in this PR?"

File: ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala (hunk 8 above)
[the remark this answers is not in the document]
Comment (tudor-da) [reply ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/committer/TransactionCommitter.scala:263]: @fabiotudone-da Sure, it is for reducing the burden on the flows downstream (i.e. we aim at increasing throughput in the reference architecture: see the ticket linked in the description). Although, just to make it clear, this is not going to be solved with this PR since it has breaking changes. This PR is just for allowing to see what the impact will be (from a performance POV) on the other components of the read path

**Comment by tudor-da** [comment]
> > I understand this is still a draft, so just some early feedback. Looks good so far.
> > For my understanding: are we planning to make this change part of the protocol between the committer and the indexer, allowing the indexer to assume there will be no unnecessary¹ nodes in the transaction? If so, do we ultimately want to also remove the filtering we're already performing there in this PR?
> > ¹ from a Ledger API perspective
> 
> In my review I left this comment but it probably went unnoticed.
> 
> tl;dr "do we ultimately want to also remove the filtering we're already performing [in the indexer] in this PR?"

Hi @stefanobaghino-da . It might be, I am not sure at the moment, maybe @gerolf-da has a more revealing input. As mentioned in a previous comment, this is a sentinel PR, it is not meant to be merged (see the ticket linked in the description). It's sole purpose is to allow us to create a custom execution engine image so we can benchmark the impact the changes have on other components.
